src/main/java/pro/trevor/tankgame/rule/impl/shared/TickRules.java

```diff
@@ -40,6 +44,17 @@ public class TickRules {
                 }
             });
 
+    public static <E extends Enum<E> & IAttribute> TickActionRule<GenericTank<E>> GetHealTanksInHealthPoolRule() {
+        return new TickActionRule<>(
+                (s, t) -> {
+                    if (t.getBoolean((E) TankAttribute.DEAD))
```

Comment:
It would probably make more sense to have string (or int) constants index the generic entity attributes. @TrevorBrunette had some concerns about type safety so maybe something like this would work?

```java
GenericElement myTank = /* ... */
bool isDead = DeadAttribute.from(myTank);

GenericElement myGoldMineTile = /* ... */
bool isDead = DeadAttribute.from(myGoldMineTime); // throws error because DeadAttribute is not defined for gold mine tile
```

This allows a lot us to define attributes that are not tied to a version that have an enforced type and don't just end up in a giant unorganized enum... thoughts?

Reply:
I was thinking we might be able to just have a `getUnsafe(String)` on `GenericElement` that attempts to cast to the given enum instance using reflection. This way we can use enums still to represent attributes where we can, but we can also get elements off of a `GenericElement` generically when necessary.

Reply:
My goal is to write all rules in terms of generic tank (benefits are sharing rules between versions and testability). So it's not going to be "when necessary", it's going to be always.

Reply:
I think fragmenting the API (`getUnsafe` to be generic and hard coded enum for static) doesn't make much sense at that point the enums are effectively string constants.  I'd be ok with a single global enum or string constants as well but we need to get away from version specific state if we're going to share rules across game versions.

Reply:
Ryan, I like your idea, but am confused about how implementation would work in the "from" function. I suppose on the client side, you would have to catch exceptions in the rules code?

Reply:
For the record, I knew this wasn't going to work, but wanted to have this conversation. We are talking about a larger redesign though, so either we push this in and fix it later, or wait for that redesign

Reply:
The from function would call `GenericElement.get` or maybe directly parse a json value.  In theory the Attribute classes would also have a to or set method to store the attributes on the `GenericElement`.  And yes calling `Attribute.from` on a tank that didn't have that attribute would throw an exception which would be caught in testing or by the client (presumably there would be an `Attribute.has` or `Attribute.optionalFrom`).

Reply:
> For the record, I knew this wasn't going to work, but wanted to have this conversation. We are talking about a larger redesign though, so either we push this in and fix it later, or wait for that redesign

I don't like giving the impression that this code works for v3+ tanks if it doesn't.  If you want to push this through now update it to use verion5 TankAttribute or version3 TankAttribute explicitly or take TankAttribute.DEAD as a parameter.

Reply:
I think the rules and tests are now decoupled from v3 vs v5 (etc.) Tanks.